File: src/kubeconfig.ts
import { homedir } from "os";
import { join } from "path";
import { readFile } from "fs/promises";
import { existsSync } from "fs";
import { writeFilePreservingMode } from "./atomic-write.js";
import log from "electron-log/main";
import { ClusterInfo } from "./aws-eks.js";
import * as yaml from "js-yaml";

const AWS_IAM_AUTHENTICATOR_BASENAME =
    process.platform === "win32"
        ? "aws-iam-authenticator.exe"
        : "aws-iam-authenticator";

/**
 * Locates the bundled authenticator that `extraResource` in forge.config.js
 * copies into the package.
 *
 * It ends up in two places: `resources/` from `extraResource` itself, and
 * `resources/app/` because the packager also copies the project directory
 * wholesale. Older builds only had the second one. The path is written into
 * the user's kubeconfig, so guessing wrong leaves an entry that fails only
 * later, when kubectl tries to run it — probe for it instead.
 */
function findAwsIamAuthenticator(): string {
    const override = process.env["AWS_IAM_AUTHENTICATOR_PATH"];
    if (override) {
        return override;
    }

    const candidates = [
        join(process.resourcesPath, AWS_IAM_AUTHENTICATOR_BASENAME),
        join(process.resourcesPath, "app", AWS_IAM_AUTHENTICATOR_BASENAME),
    ];

    const found = candidates.find((candidate) => existsSync(candidate));
    if (!found) {
        log.warn(
            "[findAwsIamAuthenticator] Not found in %s, falling back to %s",
            candidates.join(", "),
            candidates[0]
        );
    }
    return found || candidates[0];
}

const AWS_IAM_AUTHENTICATOR = findAwsIamAuthenticator();

type NamePattern = (info: ClusterInfo) => string;

/** The per-entry key that holds the body, one per kubeconfig list. */
type Section = "cluster" | "user" | "context";

/** One entry of `clusters`, `users` or `contexts`. */
interface NamedEntry {
    name?: unknown;
    [key: string]: unknown;
}

/**
 * A kubeconfig as it is on disk. Only the three lists are spelled out because
 * they are the only thing Frost writes; `current-context`, `preferences`,
 * per-entry `extensions` and whatever kubectl grows next ride along in the
 * index signature and are written back untouched.
 */
interface KubeconfigDocument {
    clusters?: NamedEntry[];
    users?: NamedEntry[];
    contexts?: NamedEntry[];
    [key: string]: unknown;
}

function kubeconfigPath(): string {
    return join(homedir(), ".kube", "config");
}

async function readKubeconfigFile(fullPath: string): Promise<string> {
    try {
        return (await readFile(fullPath)).toString();
    } catch (err) {
        log.debug(
            "[readKubeconfigFile] While trying to read %s: %s",
            fullPath,
            err
        );
        return "";
    }
}

export async function writeKubeconfig(clusters: ClusterInfo[]) {
    const fullPath = kubeconfigPath();
    const existing = await readKubeconfigFile(fullPath);

    let contents: string;
    try {
        contents = mergeKubeconfig(existing, clusters);
    } catch (err) {
        // Most of this file is the user's - their own clusters, their
        // current-context. One we cannot parse is one we cannot merge into,
        // and writing our entries on their own would throw the rest away.
        log.error(
            "[writeKubeconfig] Leaving %s alone, cannot parse it: %s",
            fullPath,
            err
        );
        return;
    }

    if (contents === existing) {
        log.info("[writeKubeconfig] %s is up to date", fullPath);
        return;
    }

    log.info("[writeKubeconfig] Writing %s", fullPath);
    // Holds the user's own clusters as well as ours, so an interrupted run
    // must not truncate it and their permissions are theirs to keep.
    await writeFilePreservingMode(fullPath, contents);
}

/**
 * Merges the discovered clusters into an existing kubeconfig.
 *
 * The parsed document is edited in place rather than rebuilt from a model of a
 * kubeconfig. A model only carries the fields it knows about, so a round trip
 * through one drops the rest - `preferences`, `extensions`, impersonation keys
 * - and a single entry it refuses to parse costs the user every context and
 * user in the file. `kubectl config set-context x --namespace=y` writes such
 * an entry: a context with no cluster.
 *
 * Pure and importable, so it can be exercised without an Electron app.
 */
export function mergeKubeconfig(
    existing: string,
    infos: ClusterInfo[]
): string {
    const document = parseKubeconfig(existing);
    const getName = getNamePattern(infos);

    for (const info of infos) {
        const name = getName(info);

        document.clusters = upsert(document.clusters, name, "cluster", {
            server: info.cluster.endpoint!,
            "certificate-authority-data":
                info.cluster.certificateAuthority!.data!,
        });

        document.users = upsert(document.users, name, "user", {
            exec: {
                apiVersion: "client.authentication.k8s.io/v1",
                command: AWS_IAM_AUTHENTICATOR,
                args: ["token", "-i", info.cluster.name!],
                env: [{ name: "AWS_PROFILE", value: info.profile.name }],
                interactiveMode: "Never",
            },
        });

        document.contexts = upsert(document.contexts, name, "context", {
            cluster: name,
            user: name,
        });
    }

    // Every value on one line. Folding a base64 certificate or an application
    // path across lines is legal YAML and kubectl reads it back, but it leaves
    // a file unlike what `aws eks update-kubeconfig` writes and unreadable in
    // a diff.
    return yaml.dump(document, { lineWidth: -1 });
}

function parseKubeconfig(existing: string): KubeconfigDocument {
    if (existing.trim() === "") {
        return { apiVersion: "v1", kind: "Config", preferences: {} };
    }

    const document = yaml.load(existing);
    if (
        typeof document !== "object" ||
        document === null ||
        Array.isArray(document)
    ) {
        throw new Error("kubeconfig is not a YAML mapping");
    }
    return document as KubeconfigDocument;
}

/**
 * Writes one `{ name, <section>: {...} }` entry, replacing the same-named one
 * if it is already there.
 *
 * That replacement merges, and keeps the entry where it was: Frost owns the
 * keys it writes and nothing else, so a namespace the user set on one of these
 * contexts, or a `proxy-url` on the cluster, survives the next refresh.
 */
function upsert(
    entries: NamedEntry[] | undefined,
    name: string,
    section: Section,
    body: Record<string, unknown>
): NamedEntry[] {
    const list = Array.isArray(entries) ? entries : [];
    const index = list.findIndex((entry) => entry?.name === name);

    if (index < 0) {
        return [...list, { name, [section]: body }];
    }

    const previous = list[index][section];
    const merged =
        typeof previous === "object" && previous !== null
            ? { ...previous, ...body }
            : body;

    return list.map((entry, at) =>
        at === index ? { ...entry, name, [section]: merged } : entry
    );
}

function getNamePattern(infos: ClusterInfo[]): NamePattern {
    const clusterNames = infos.map((info) => info.cluster.name);
    const clusterIds = infos.map(
        (info) =>
            `${info.cluster.name}:${info.profile.accountName}:${info.region.RegionName}`
    );
    const roleNames = infos.map((info) => info.profile.roleName);
    const regionNames = infos.map((info) => info.region.RegionName);

    // Whether the bare name tells two *different* clusters apart. One cluster
    // reached through several profiles is one id and one name; two clusters
    // that happen to share a name are two ids and one name, and every pattern
    // below would then hand them the same entry and lose one of them.
    const uniqueClusters =
        new Set(clusterNames).size === new Set(clusterIds).size;
    const sameRoleName = new Set(roleNames).size === 1;
    const sameRegion = new Set(regionNames).size === 1;

    if (uniqueClusters) {
        if (sameRoleName && sameRegion) {
            return (info: ClusterInfo) => `${info.cluster.name}`;
        }
        if (sameRoleName) {
            return (info: ClusterInfo) =>
                `${info.cluster.name}:${info.region.RegionName}`;
        }
        if (sameRegion) {
            return (info: ClusterInfo) =>
                `${info.cluster.name}:${info.profile.roleName}`;
        }
        return (info: ClusterInfo) =>
            `${info.cluster.name}:${info.region.RegionName}:${info.profile.roleName}`;
    }

    return (info: ClusterInfo) =>
        `${info.cluster.name}:${info.profile.accountName}:${info.region.RegionName}:${info.profile.roleName}`;
}

File: frost-core/tests/kubeconfig.rs
//! Integration tests for kubeconfig context-name selection and merging.

use frost_core::kubeconfig::{
    ClusterInfo, KubeConfig, NamePattern, choose_name_pattern, merge_kubeconfig,
};

fn info(cluster: &str, account: &str, role: &str, region: &str) -> ClusterInfo {
    ClusterInfo {
        cluster_name: cluster.into(),
        account_name: account.into(),
        role_name: role.into(),
        region: region.into(),
        profile_name: format!("{account}-{role}"),
        endpoint: "https://eks.example".into(),
        ca_data: "Y2E=".into(),
    }
}

#[test]
fn single_cluster_uses_bare_name() {
    let infos = vec![info("prod", "acme", "admin", "us-east-1")];
    assert_eq!(choose_name_pattern(&infos), NamePattern::ClusterOnly);
    assert_eq!(choose_name_pattern(&infos).format(&infos[0]), "prod");
}

#[test]
fn same_role_different_region_appends_region() {
    let infos = vec![
        info("prod", "acme", "admin", "us-east-1"),
        info("stage", "acme", "admin", "eu-west-1"),
    ];
    assert_eq!(choose_name_pattern(&infos), NamePattern::ClusterRegion);
    assert_eq!(
        choose_name_pattern(&infos).format(&infos[0]),
        "prod:us-east-1"
    );
}

#[test]
fn same_region_different_role_appends_role() {
    let infos = vec![
        info("prod", "acme", "admin", "us-east-1"),
        info("stage", "acme", "poweruser", "us-east-1"),
    ];
    assert_eq!(choose_name_pattern(&infos), NamePattern::ClusterRole);
    assert_eq!(
        choose_name_pattern(&infos).format(&infos[1]),
        "stage:poweruser"
    );
}

#[test]
fn differing_role_and_region_appends_both() {
    let infos = vec![
        info("prod", "acme", "admin", "us-east-1"),
        info("stage", "acme", "poweruser", "eu-west-1"),
    ];
    assert_eq!(choose_name_pattern(&infos), NamePattern::ClusterRegionRole);
    assert_eq!(
        choose_name_pattern(&infos).format(&infos[0]),
        "prod:us-east-1:admin"
    );
}

#[test]
fn colliding_cluster_names_use_full_pattern() {
    // Same cluster name reached via two accounts must not collide.
    let infos = vec![
        info("shared", "acme", "admin", "us-east-1"),
        info("shared", "beta", "admin", "us-east-1"),
    ];
    assert_eq!(choose_name_pattern(&infos), NamePattern::Full);
    assert_eq!(
        choose_name_pattern(&infos).format(&infos[0]),
        "shared:acme:us-east-1:admin"
    );
    assert_eq!(
        choose_name_pattern(&infos).format(&infos[1]),
        "shared:beta:us-east-1:admin"
    );
}

#[test]
fn merge_into_empty_produces_valid_config() {
    let infos = vec![info("prod", "acme", "admin", "us-east-1")];
    let yaml = merge_kubeconfig(None, &infos, "/path/aws-iam-authenticator").unwrap();
    let parsed: KubeConfig = serde_yaml::from_str(&yaml).unwrap();
    assert_eq!(parsed.api_version, "v1");
    assert_eq!(parsed.kind, "Config");
    assert_eq!(parsed.clusters.len(), 1);
    assert_eq!(parsed.clusters[0].name, "prod");
    assert_eq!(parsed.users[0].user.exec.args, vec!["token", "-i", "prod"]);
    assert_eq!(parsed.users[0].user.exec.env[0].value, "acme-admin");
    assert_eq!(parsed.contexts[0].context.cluster, "prod");
}

#[test]
fn merge_replaces_same_name_and_keeps_others() {
    let existing = "\
apiVersion: v1
kind: Config
clusters:
- name: other
  cluster:
    server: https://other
    certificate-authority-data: eC0=
- name: prod
  cluster:
    server: https://stale
    certificate-authority-data: c3RhbGU=
users: []
contexts: []
";
    let infos = vec![info("prod", "acme", "admin", "us-east-1")];
    let yaml = merge_kubeconfig(Some(existing), &infos, "/auth").unwrap();
    let parsed: KubeConfig = serde_yaml::from_str(&yaml).unwrap();

    let names: Vec<&str> = parsed.clusters.iter().map(|c| c.name.as_str()).collect();
    assert!(names.contains(&"other"));
    assert!(names.contains(&"prod"));
    assert_eq!(parsed.clusters.len(), 2);
    let prod = parsed.clusters.iter().find(|c| c.name == "prod").unwrap();
    assert_eq!(prod.cluster.server, "https://eks.example");
}
